lib/annotations2triannon/open_annotation.rb

```diff
@@ -261,6 +254,33 @@ def to_ttl
       @graph.dump(:ttl, standard_prefixes: true)
     end
 
+    private
+
+    # Sets the annotation ID as an RDF::URI from id, but if the id is nil, it
+    # will try to get the ID from the graph and, as a last resort, it will
+    # generate a unique ID using a UUID.
+    # @param id [UUID|URI|String] to identify an open annotation
+    def set_id(id)
+      raise ArgumentError, 'id cannot be an empty String' if (id.instance_of?(String) && id.empty?)
+      if id.nil?
+        # Try to get the ID from the graph or use a UUID
+        q = [nil, RDF.type, OA.Annotation]
+        @id = @graph.query(q).collect {|s| s.subject }.first || RDF::URI.parse(UUID.generate)
```

Comment:
I'm not an authority on this much debated topic, but I fall on the side of the fence that wants to avoid blank nodes.  For discussion, e.g.
  - http://milicicvuk.com/blog/2011/07/14/problems-of-the-rdf-model-blank-nodes/